// react-bootstrap-editable/dist/editable.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var PropTypes = _interopDefault(require('prop-types'));
var reactstrap = require('reactstrap');

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _isNativeReflectConstruct() {
  if (typeof Reflect === "undefined" || !Reflect.construct) return false;
  if (Reflect.construct.sham) return false;
  if (typeof Proxy === "function") return true;

  try {
    Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
    return true;
  } catch (e) {
    return false;
  }
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _createSuper(Derived) {
  var hasNativeReflectConstruct = _isNativeReflectConstruct();

  return function _createSuperInternal() {
    var Super = _getPrototypeOf(Derived),
        result;

    if (hasNativeReflectConstruct) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var TextField = /*#__PURE__*/function (_React$Component) {
  _inherits(TextField, _React$Component);

  var _super = _createSuper(TextField);

  function TextField(props) {
    _classCallCheck(this, TextField);

    return _super.call(this, props);
  }

  _createClass(TextField, [{
    key: "render",
    value: function render() {
      var _this = this;

      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(reactstrap.Input, {
        invalid: !!this.props.validationText,
        autoFocus: true,
        value: this.props.value ? this.props.value : "",
        onChange: function onChange(e) {
          return _this.props.setNewValue(e.target.value);
        },
        type: "text",
        bsSize: "sm",
        className: "mr-1"
      }), this.props.controls);
    }
  }]);

  return TextField;
}(React.Component);

var TextArea = /*#__PURE__*/function (_React$Component) {
  _inherits(TextArea, _React$Component);

  var _super = _createSuper(TextArea);

  function TextArea(props) {
    _classCallCheck(this, TextArea);

    return _super.call(this, props);
  }

  _createClass(TextArea, [{
    key: "render",
    value: function render() {
      var _this = this;

      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(reactstrap.Input, {
        invalid: !!this.props.validationText,
        autoFocus: true,
        value: this.props.value ? this.props.value : "",
        onChange: function onChange(e) {
          return _this.props.setNewValue(e.target.value);
        },
        type: "textarea",
        className: "mb-1"
      }));
    }
  }]);

  return TextArea;
}(React.Component);

var TextField$1 = /*#__PURE__*/function (_React$Component) {
  _inherits(TextField, _React$Component);

  var _super = _createSuper(TextField);

  function TextField(props) {
    _classCallCheck(this, TextField);

    return _super.call(this, props);
  }

  _createClass(TextField, [{
    key: "render",
    value: function render() {
      var _this = this;

      var options = this.props.options.map(function (value, index) {
        return /*#__PURE__*/React.createElement("option", {
          key: index + value
        }, value);
      });
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(reactstrap.Input, {
        value: this.props.value,
        onChange: function onChange(e) {
          return _this.props.setNewValue(e.target.value);
        },
        type: "select",
        bsSize: "sm",
        className: "mr-1"
      }, options), this.props.controls);
    }
  }]);

  return TextField;
}(React.Component);

var Date$1 = /*#__PURE__*/function (_React$Component) {
  _inherits(Date, _React$Component);

  var _super = _createSuper(Date);

  function Date(props) {
    _classCallCheck(this, Date);

    return _super.call(this, props);
  }

  _createClass(Date, [{
    key: "render",
    value: function render() {
      var _this = this;

      var date = this.props.value;

      if (date) {
        date = date.toISOString ? date.toISOString().slice(0, 10) : new window.Date(date).toISOString().slice(0, 10);
      }

      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(reactstrap.Input, {
        value: date ? date : "",
        type: "date",
        bsSize: "sm",
        className: "mr-1",
        onChange: function onChange(e) {
          return _this.props.setNewValue(e.target.valueAsDate);
        }
      }), this.props.controls);
    }
  }]);

  return Date;
}(React.Component);

var File = /*#__PURE__*/function (_React$Component) {
  _inherits(File, _React$Component);

  var _super = _createSuper(File);

  function File(props) {
    _classCallCheck(this, File);

    return _super.call(this, props);
  }

  _createClass(File, [{
    key: "render",
    value: function render() {
      var _this = this;

      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(reactstrap.CustomInput, {
        className: "form-control-sm mr-1",
        type: "file",
        bsSize: "sm",
        label: this.props.value ? this.props.value.name : this.props.label,
        onChange: function onChange(e) {
          return _this.props.setNewValue(e.target.files[0]);
        }
      }), this.props.controls);
    }
  }]);

  return File;
}(React.Component);

var fontAwesomeStyle = {
  textAlign: "center",
  width: "1.25em",
  height: "1em",
  fill: "white",
  verticalAlign: "-0.25em"
};

var Editable = /*#__PURE__*/function (_React$Component) {
  _inherits(Editable, _React$Component);

  var _super = _createSuper(Editable);

  function Editable(props) {
    var _this;

    _classCallCheck(this, Editable);

    _this = _super.call(this, props);
    _this.state = {
      value: _this.props.initialValue,
      newValue: _this.props.initialValue,
      isEditing: false,
      validationText: null,
      isLoading: false
    }; //used for popover mode

    _this.clickableLink = /*#__PURE__*/React.createRef();
    return _this;
  }

  _createClass(Editable, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.ajax && !this.props.validate && !this.props.disabled) {
        console.error("Editable(".concat(this.props.id, "): You provided an ajax prop without a validate prop;\n            ajax function will not be called"));
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      //update initial value if the prop got updated
      if (prevProps.initialValue !== this.props.initialValue) {
        this.setState({
          value: this.props.initialValue,
          newValue: this.props.initialValue
        });
      }
    }
  }, {
    key: "getEditingComponent",
    value: function getEditingComponent() {
      var _this2 = this;

      var confirmButton = /*#__PURE__*/React.createElement(reactstrap.Button, {
        className: "ml-auto mr-1",
        color: "success",
        size: "sm"
      }, /*#__PURE__*/React.createElement("svg", {
        xmlns: "http://www.w3.org/2000/svg",
        viewBox: "0 0 512 512",
        style: fontAwesomeStyle
      }, /*#__PURE__*/React.createElement("path", {
        color: "white",
        d: "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
      })));
      var cancelButton = /*#__PURE__*/React.createElement(reactstrap.Button, {
        color: "danger",
        size: "sm",
        onClick: function onClick() {
          return _this2.onCancel();
        }
      }, /*#__PURE__*/React.createElement("svg", {
        xmlns: "http://www.w3.org/2000/svg",
        viewBox: "0 0 352 512",
        style: fontAwesomeStyle
      }, /*#__PURE__*/React.createElement("path", {
        d: "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
      })));

      if ( /*#__PURE__*/React.isValidElement(this.props.renderConfirmElement)) {
        confirmButton = /*#__PURE__*/React.cloneElement(this.props.renderConfirmElement, {
          onClick: function onClick(e) {
            return _this2.onFormSubmit(e);
          }
        });
      }

      if ( /*#__PURE__*/React.isValidElement(this.props.renderCancelElement)) {
        cancelButton = /*#__PURE__*/React.cloneElement(this.props.renderCancelElement, {
          onClick: function onClick(e) {
            e.preventDefault();

            _this2.onCancel();
          }
        });
      }

      var controls = /*#__PURE__*/React.createElement(React.Fragment, null, confirmButton, cancelButton);

      if (this.state.isLoading) {
        controls = /*#__PURE__*/React.createElement("div", {
          className: "my-auto mx-4"
        }, /*#__PURE__*/React.createElement(reactstrap.Spinner, {
          style: {
            width: "1.5rem",
            height: "1.5rem"
          }
        }));
      }

      var commonProps = {
        value: this.state.newValue,
        validationText: this.state.validationText,
        controls: controls,
        setNewValue: function setNewValue(newValue) {
          return _this2.setState({
            newValue: newValue
          });
        },
        onCancel: function onCancel() {
          return _this2.onCancel();
        }
      };
      var component;

      switch (this.props.type) {
        case "textfield":
          component = /*#__PURE__*/React.createElement(TextField, commonProps);
          break;

        case "select":
          component = /*#__PURE__*/React.createElement(TextField$1, _extends({}, commonProps, {
            options: this.props.options
          }));
          break;

        case "date":
          component = /*#__PURE__*/React.createElement(Date$1, commonProps);
          break;

        case "textarea":
          return /*#__PURE__*/React.createElement(reactstrap.Form, {
            onSubmit: function onSubmit(e) {
              return _this2.onFormSubmit(e);
            }
          }, /*#__PURE__*/React.createElement(TextArea, commonProps), /*#__PURE__*/React.createElement("div", {
            className: "d-flex align-items-start"
          }, /*#__PURE__*/React.createElement(reactstrap.FormText, {
            className: "mt-0"
          }, this.state.validationText), controls));

        case "file":
          component = /*#__PURE__*/React.createElement(File, _extends({}, commonProps, {
            label: this.props.label
          }));
          break;

        default:
          console.error("Editable(".concat(this.props.id, "): \"").concat(this.props.type, "\" is not a valid value for the \"type\" prop"));
          return null;
      }

      return /*#__PURE__*/React.createElement(reactstrap.Form, {
        onSubmit: function onSubmit(e) {
          return _this2.onFormSubmit(e);
        },
        className: this.props.className
      }, /*#__PURE__*/React.createElement("div", {
        className: "align-items-baseline d-flex"
      }, component), /*#__PURE__*/React.createElement(reactstrap.FormText, {
        className: "mt-0"
      }, this.state.validationText));
    }
  }, {
    key: "onFormSubmit",
    value: function onFormSubmit(e) {
      e.preventDefault();
      this.onSubmit(this.state.newValue);
    }
  }, {
    key: "onCancel",
    value: function onCancel() {
      //reset validation text AND new value, all back to initial
      this.setState({
        validationText: null,
        newValue: this.state.value,
        isEditing: false
      });
    } //validation happens here

  }, {
    key: "onSubmit",
    value: function onSubmit(newValue) {
      var validationText = this.props.validate ? this.props.validate(newValue) : null; //we always trigger this, as long as the prop is specified

      this.props.onSubmit ? this.props.onSubmit(newValue) : null;

      if (validationText) {
        this.setState({
          validationText: validationText
        });
      } else {
        this.props.validate ? this.onValidated(newValue) : this.setState({
          value: newValue,
          isEditing: false
        });
      }
    }
  }, {
    key: "onValidated",
    value: function onValidated(validValue) {
      if (this.props.onValidated) {
        this.props.onValidated(validValue);
      } else if (!this.props.ajax) {
        console.error("Editable(".concat(this.props.id, "): Specified a validate function without onValidated or ajax"));
      }

      if (this.props.ajax && validValue !== this.state.value) {
        this.ajax(validValue);
      } else {
        this.setState({
          value: validValue,
          isEditing: false,
          validationText: null
        });
      }
    }
  }, {
    key: "ajax",
    value: function ajax(validValue) {
      var _this3 = this;

      this.setState({
        isLoading: true
      });
      var xhr = new XMLHttpRequest(); //this will call the user's ajax function, allowing him to set up the xhr object however he wants

      this.props.ajax(xhr, validValue, this.props.id); //consume the user's on ready state change function to call it later before the editable's

      var onReadyStateChange = xhr.onreadystatechange ? xhr.onreadystatechange : null;

      xhr.onreadystatechange = function () {
        onReadyStateChange ? onReadyStateChange() : null;

        if (xhr.readyState === 4) {
          if (xhr.status === 200) {
            _this3.setState({
              isLoading: false,
              isEditing: false,
              value: validValue,
              validationText: null
            });
          } else {
            _this3.setState({
              isLoading: false,
              validationText: "Ajax Response ".concat(xhr.status, " Error")
            });
          }
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      if ((this.state.isEditing || this.props.alwaysEditing) && this.props.mode === "inline") {
        return this.getEditingComponent();
      } else {
        var value = this.state.value ? this.state.value : "No value"; //format date objects for display, might add a custom format function here later

        value = this.props.type === "date" && this.state.value ? new window.Date(this.state.value).toUTCString().slice(5, 16) : value;
        value = this.props.type === "file" && this.state.value ? this.state.value.name : value;
        var p = "",
            a = "";

        if (this.props.isValueClickable) {
          if (this.props.disabled) {
            p = value;
          } else {
            a = value;
          }
        } else {
          p = value;
          a = this.props.disabled ? a : this.props.editText;
        } //add label if applicable


        p = this.props.label ? "".concat(this.props.label, ": ").concat(p) : p;
        var popover = this.props.mode === "popover" ? /*#__PURE__*/React.createElement(reactstrap.Popover, {
          isOpen: this.state.isEditing,
          placement: this.props.placement,
          target: this.clickableLink
        }, /*#__PURE__*/React.createElement(reactstrap.PopoverHeader, null, this.props.label), /*#__PURE__*/React.createElement(reactstrap.PopoverBody, null, this.getEditingComponent())) : null;
        return /*#__PURE__*/React.createElement(reactstrap.Form, {
          onSubmit: function onSubmit(e) {
            return _this4.onFormSubmit(e);
          },
          className: this.props.className,
          inline: true
        }, p && this.props.showText && /*#__PURE__*/React.createElement("p", {
          className: "my-0",
          style: {
            "whiteSpace": "pre-wrap"
          }
        }, p), a && /*#__PURE__*/React.createElement("a", {
          ref: this.clickableLink,
          className: "ml-1 mt-auto",
          href: "#",
          onClick: function onClick(e) {
            e.preventDefault();

            _this4.setState({
              isEditing: true
            });
          }
        }, a), popover);
      }
    }
  }]);

  return Editable;
}(React.Component);
Editable.defaultProps = {
  type: "textfield",
  mode: "inline",
  alwaysEditing: false,
  className: null,
  initialValue: null,
  id: null,
  label: null,
  showText: true,
  disabled: false,
  isValueClickable: false,
  editText: "Edit",
  renderConfirmElement: null,
  renderCancelElement: null,
  //popover
  placement: "top",
  //functions
  validate: null,
  ajax: null,
  onSubmit: null,
  onValidated: null,
  //select props
  options: null
};
Editable.propTypes = {
  type: PropTypes.oneOf(["textfield", "textarea", "select", "date", "file"]).isRequired,
  mode: PropTypes.oneOf(["inline", "popover"]).isRequired,
  alwaysEditing: PropTypes.bool,
  className: PropTypes.string,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date$1)]),
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  showText: PropTypes.bool,
  disabled: PropTypes.bool,
  isValueClickable: PropTypes.bool,
  editText: PropTypes.string,
  renderConfirmElement: PropTypes.element,
  renderCancelElement: PropTypes.element,

  /** Functions */
  validate: PropTypes.func,
  ajax: PropTypes.func,
  onSubmit: PropTypes.func,
  onValidated: PropTypes.func,

  /** Popover mode only */
  placement: PropTypes.oneOf(["auto", "auto-start", "auto-end", "top", "top-start", "top-end", "right", "right-start", "right-end", "bottom", "bottom-start", "bottom-end", "left", "left-start", "left-end"]),

  /** Select only */
  options: PropTypes.array
};

module.exports = Editable;
